File: modules/Windows/notificationPopups.ts
import { Notification } from "../Widgets/index"
import options from "options"
import { Utils, Notifications, Widget, Service } from "imports"

const notifications = await Service.import("notifications")
const { transition } = options
const { position } = options.notifications
const { timeout, idle } = Utils
const { Box, Label } = Widget

function Animated(id: number) {
    const n = notifications.getNotification(id)!
    const widget = Notification(n)

    const inner = Widget.Revealer({
        transition: "slide_up",
        transition_duration: transition.value,
        child: widget,
    })

    const outer = Widget.Revealer({
        transition: "slide_down",
        transition_duration: transition.value,
        child: inner,
    })

    const box = Box({
        hpack: "end",
        child: outer,
    })

    idle(() => {
        outer.reveal_child = true
        timeout(transition.value, () => {
            inner.reveal_child = true
        })
    })

    return Object.assign(box, {
        dismiss() {
            inner.reveal_child = false
            timeout(transition.value, () => {
                outer.reveal_child = false
                timeout(300, () => {
                    box.destroy()
                })
            })
        },
    })
}

function PopupList() {
    const map: Map<number, ReturnType<typeof Animated>> = new Map
    const box = Box({
        vertical: true,
        hexpand: true,
        hpack: "fill",
        css: options.notifications.width.bind().as(w => `min-width: ${w}px;`),
    })

    function remove(_: unknown, id: number) {
        map.get(id)?.dismiss()
        map.delete(id)
    }

    return box
        .hook(notifications, (_, id: number) => {
            if (id !== undefined) {
                if (map.has(id))
                    remove(null, id)

                if (notifications.dnd)
                    return

                const w = Animated(id)
                map.set(id, w)
                box.children = [w, ...box.children]
            }
        }, "notified")
        .hook(notifications, remove, "dismissed")
        .hook(notifications, remove, "closed")
}

export default (monitor: number) => Widget.Window({
    monitor,
    name: `notifications${monitor}`,
    anchor: position.bind(),
    className: "notifications",
    hexpand: true,
    child: Box({
        css: "padding: 2px",
        child: PopupList(),
    }),
})

File: Rust/Modules/Windows/NotificationPopups.rs
import std::collections::HashMap;

use crate::Widgets::Notification;
use crate::options;
use crate::Utils::{idle, timeout};
use crate::imports::{Notifications, Service, Widget};

fn animated(id: i32, notifications: &Notifications, transition: &str) -> Widget {
    let n = notifications.get_notification(id).unwrap();
    let widget = Notification(n);

    let inner = Widget::Revealer {
        transition: "slide_up".to_string(),
        transition_duration: options::transition().value,
        child: widget,
    };

    let outer = Widget::Revealer {
        transition: "slide_down".to_string(),
        transition_duration: options::transition().value,
        child: inner,
    };

    let box = Widget::Box {
        hpack: "end".to_string(),
        child: outer,
    };

    idle(|| {
        outer.reveal_child = true;
        timeout(options::transition().value, || {
            inner.reveal_child = true;
        });
    });

    box
}

fn popup_list(notifications: &Notifications, options: &options) -> Widget {
    let mut map: HashMap<i32, Widget> = HashMap::new();
    let box = Widget::Box {
        vertical: true,
        hexpand: true,
        hpack: "fill".to_string(),
        css: options.notifications.width().bind().as(|w| format!("min-width: {}px;", w)),
    };

    fn remove(map: &mut HashMap<i32, Widget>, id: i32) {
        if let Some(w) = map.get(&id) {
            w.dismiss();
            map.remove(&id);
        }
    }

    box.hook(notifications, |_, id| {
        if let Some(id) = id {
            if map.contains_key(&id) {
                remove(&mut map, id);
            }

            if notifications.dnd {
                return;
            }

            let w = animated(id, notifications, "slide_up");
            map.insert(id, w);
            box.children = vec![w, box.children].concat();
        }
    }, "notified")
    .hook(notifications, |_, id| remove(&mut map, id), "dismissed")
    .hook(notifications, |_, id| remove(&mut map, id), "closed");

    box
}

pub fn create_notification_window(monitor: i32) -> Widget {
    Widget::Window {
        monitor,
        name: format!("notifications{}", monitor),
        anchor: options::position().bind(),
        class_name: "notifications".to_string(),
        hexpand: true,
        child: Widget::Box {
            css: "padding: 2px".to_string(),
            child: popup_list(&Notifications::import("notifications"), &options),
        },
    }
}